### DISCUSSION TITLE
Is NewPipe internally sabotaged by undercover Google devs?

### DISCUSSION THREAD

**Discussion by highbaser** [TeamNewPipe/NewPipe#11005] [Questions]
Yes, this is intentionally a conspiracy-theory-sounding, provocative question.

NewPipe and all other applications that use NewPipeExtractor are suffering from the "The video returned by YouTube isn't the requested one". One of these apps is Invidious.

See: https://github.com/iv-org/invidious/issues/4584

This issue is known for at least a month and nothing seems to have been done about it until it rendered NewPipeExtractor completely unusable. So basically, while current releases of NewPipe and NewPipeExtractor are getting more and more unusable, you're working on refactors that could've waited. Also, NewPipe is also almost at its knees. It constantly reporting "Something went wrong" during normal usage. It doesn't load comments for any video and going down the same unusability path as apps using NewPipeExtractor. This raises questions.

Why don't you prioritize stability and usability of NewPipe over refactoring needs?

Why have you scheduled your great refactors to times when Youtube is also hunting down 3rd-party apps (ad blockers, 3rd-party clients, etc.)?

**Comment by iTrooz** [comment]
Hi, I'm not part of the team, but I think I can add some insight to this.

I'll assume you are a user, and not a developer.

In general, refactors are needed because when you build a large codebase (or any software really), most of the time you will accumulate a "technical debt" over time.

That technical debt essentially means it gets harder to develop new features/fix bugs/nagivate the codebase. And it makes it less enjoyable to work on the software (fighting against the codebase is NOT fun).

Refactors help to decease the technical debt, and establish ways to work with it more efficiently

As to why do it now, when Youtube began to ban third-party clients ? Way more probably to be just a coincidence (+, the refactors started before youtube began it's third party client war IIRC)

TLDR; in general refactors are a way to plan for long-term rather than short-term fixes.
This message does not replace insight from the team on this particular project

**Reply by highbaser** [reply]
> In general, refactors are needed because when you build a large codebase (or any software really), most of the time you will accumulate a "technical debt" over time.

I was not questioning the necessity of refactors. I was questioning the **timing**. Refactors are generally don't have as high priority as major bugs do. Major bugs include those which render the software unusable or near-unusable, like the one I mentioned. Refactors are rather a long term technical debt.

> Way more probably to be just a coincidence

That's what I'm not exactly sure about. :)

**Reply by TobiGr** [reply]
The refactors have been planned a long time ago. See https://github.com/TeamNewPipe/NewPipe/discussions/10118
There is nothing to worry about.

**Reply by bohrium2b** [reply]
Maybe you've seen the memes, but in general, when a refactor is needed, a refactor is needed. Fixing bugs is hard when every change breaks 5 other things. I don't actually know what is happening, but timing is usually not under the control of the devs. 

> Refactors are rather a long term technical debt.

These usually have symptoms (aka. one change breaks everything else), and instead of playing whack-a-mole with a codebase, it can usually be easier to just refactor/rewrite the whole thing.

Hopefully this answers your concerns.

**Reply by fillwithjoy1** [reply]
> I was not questioning the necessity of refactors. I was questioning the **timing**. Refactors are generally don't have as high priority as major bugs do. Major bugs include those which render the software unusable or near-unusable, like the one I mentioned. Refactors are rather a long term technical debt.

Refactors are _very important_ when trying to maintain code, as they **reduce time to make new code**, **improve readability of code** and overall make development easier. As bohrium2b states, sometimes when there's too much "technical debt" it becomes near impossible to make a new change without something else breaking. This is probably the state that the NewPipe is currently in, and so they want to solidify the code they already have to stop something else from breaking (aka improving reliability)

**Comment by opusforlife2** [comment]
MWAHAHAHAHA. I'm actually Sundar Pichai in disguise, and this is a hostile takeover. Hands up! You're surrounded by Google devs! Slowly take out your phone where we can see it, and uninstall all Youtube alternatives you have.

You do _not_ have the right to remain silent. Anything you say can, and will, be recorded by Google Assistant, and then used to improve our AI datasets. We might also play it back just to embarrass you socially.

**Comment by Stypox** [comment]
Invidious is not even based on NewPipeExtractor
